#**LAB-04 : IMPLEMENTATION OF BACKPROPAGATION**

In [ ]:
import numpy as np

# Activation function: Sigmoid
def sigmoid(x):
    return 1 / (1 + np.exp(-x))

# Derivative of the sigmoid function
def sigmoid_derivative(x):
    return x * (1 - x)

# Training data (XOR problem)
inputs = np.array([[0, 0],
                   [0, 1],
                   [1, 0],
                   [1, 1]])

outputs = np.array([[0], [1], [1], [0]])

# Set random seed for reproducibility
np.random.seed(42)

# Initialize weights randomly with mean 0
input_layer_neurons = 2
hidden_layer_neurons = 2
output_neurons = 1

weights_input_hidden = np.random.uniform(size=(input_layer_neurons, hidden_layer_neurons))
weights_hidden_output = np.random.uniform(size=(hidden_layer_neurons, output_neurons))

# Learning rate
learning_rate = 0.5

# Number of iterations
epochs = 10000

# Training process
for epoch in range(epochs):
    # Forward propagation
    hidden_layer_activation = np.dot(inputs, weights_input_hidden)
    hidden_layer_output = sigmoid(hidden_layer_activation)

    final_layer_activation = np.dot(hidden_layer_output, weights_hidden_output)
    predicted_output = sigmoid(final_layer_activation)

    # Compute the error
    error = outputs - predicted_output

    # Backpropagation
    d_predicted_output = error * sigmoid_derivative(predicted_output)

    error_hidden_layer = d_predicted_output.dot(weights_hidden_output.T)
    d_hidden_layer = error_hidden_layer * sigmoid_derivative(hidden_layer_output)

    # Updating the weights
    weights_hidden_output += hidden_layer_output.T.dot(d_predicted_output) * learning_rate
    weights_input_hidden += inputs.T.dot(d_hidden_layer) * learning_rate

    if (epoch + 1) % 1000 == 0:
        loss = np.mean(np.abs(error))
        print(f'Epoch {epoch + 1}, Loss: {loss}')

# Testing the trained neural network
print("Final weights between input and hidden layers: \n", weights_input_hidden)
print("Final weights between hidden and output layers: \n", weights_hidden_output)
print("Predicted outputs after training: \n", predicted_output)

Epoch 1000, Loss: 0.39557039180027914
Epoch 2000, Loss: 0.268978650120054
Epoch 3000, Loss: 0.20877501367079165
Epoch 4000, Loss: 0.17388635703099511
Epoch 5000, Loss: 0.15095324126862225
Epoch 6000, Loss: 0.13460178571028433
Epoch 7000, Loss: 0.12227213430243797
Epoch 8000, Loss: 0.11258998461667526
Epoch 9000, Loss: 0.10475002443046719
Epoch 10000, Loss: 0.09824792584007563
Final weights between input and hidden layers: 
 [[0.90571962 7.32475086]
 [0.90572643 7.32731531]]
Final weights between hidden and output layers: 
 [[-27.46354604]
 [ 21.74964594]]
Predicted outputs after training: 
 [[0.05432736]
 [0.89823995]
 [0.89824011]
 [0.13514441]]


**Inference:**

The model's performance improved significantly as the loss decreased from 0.3956 to 0.0982 over 10000 epochs. The final weights between layers show how the inputs are combined and transformed. The predicted outputs indicate that the model can effectively distinguish between different input patterns, demonstrating its ability to make accurate predictions.